fms/modules/attention.py

```diff
@@ -174,8 +194,14 @@ def _sdpa_store_op(
     values = values.transpose(2, 1)
 
     if key_cache is not None and value_cache is not None and value_cache.numel() > 0:
-        key_cache_result = torch.cat((key_cache, keys), dim=2)
-        value_cache_result = torch.cat((value_cache, values), dim=2)
+        sliding_window = attn_kwargs.get("sliding_window", 0)
+        if sliding_window != 0:
+            sliding_window *= -1
+            sliding_window += 1
+        key_cache_result = torch.cat((key_cache[:, :, sliding_window:, :], keys), dim=2)
+        value_cache_result = torch.cat(
+            (value_cache[:, :, sliding_window:, :], values), dim=2
+        )
```

Comment:
while this is correct for decoding, given this has to work for both decoding and other things such as prefill, it might be worth making the +=1 be instead += keys.shape[x] where x is the sequence length of the keys you're appending